include/chainbase/chainrocks.hpp

```diff
@@ -0,0 +1,822 @@
+      ///   10. Operation 5: (squash)                                         ///   10. Operation 5: (squash)
+      ///   11. State 5:     S = {}, U = [{}]                                 ///   11. State 5:     S = {}, U = [{modifed,K,V}]
+      ///   12. Operation 6: (undo)                                           ///   12. Operation 6: (undo)
+      ///   13. State 6:     S = {}, U = []                                   ///   13. State 6:     S = {(K,V)}, U = []
+      ///                                                                     ///
+      /// Case 9:                                                             /// Case 9':
+      ///   1.  State 0:     S = {}, U = []                                   ///   1.  State 0:     S = {(K,V)}, U = []
+      ///   2.  Operation 1: (start_undo_session)                             ///   2.  Operation 1: (start_undo_session)
+      ///   3.  State 1:     S = {}, U = [{}]                                 ///   3.  State 1:     S = {(K,V)}, U = [{}]
+      ///   4.  Operation 2: (remove,K)                                       ///   4.  Operation 2: (remove,K)
+      ///   5.  State 2:     IMPOSSIBLE                                       ///   5.  State 2:     S = {}, U = [{(removed,K,V)}]
+      ///   6.  Operation 3: IMPOSSIBLE                                       ///   6.  Operation 3: (start_undo_session)
+      ///   7.  State 3:     IMPOSSIBLE                                       ///   7.  State 3:     S = {}, U = [{(removed,K,V)}, {}]
+      ///   8.  Operation 4: IMPOSSIBLE                                       ///   8.  Operation 4: (put,K,V')
+      ///   9.  State 4:     IMPOSSIBLE                                       ///   9.  State 4:     S = {(K,V')}, U = [{(removed,K,V)}, {(new,K)}]
+      ///   10. Operation 5: IMPOSSIBLE                                       ///   10. Operation 5: (squash)
+      ///   11. State 5:     IMPOSSIBLE                                       ///   11. State 5:     S = {(K,V')}, U = [{(modifed,K,V)}]
+      ///   12. Operation 6: IMPOSSIBLE                                       ///   12. Operation 6: (undo)
+      ///   13. State 6:     IMPOSSIBLE                                       ///   13. State 6:     S = {(K,V)}, U = []
+
+      /// Sqaushing is the act of taking the first two recent
+      /// `undo_state`s and combining them into one.
+      void squash() {
+         if(!_enabled()) {
+            return;
+         }
+
+         if(_stack.size() == 1) {
+            _stack.pop_front();
+            return;
+         }
+
+         auto& head_minus_one = _stack[_stack.size()-2];
+
+         _squash_new_keys(std::move(_stack.back()), head_minus_one);
+         _squash_modified_values(std::move(_stack.back()), head_minus_one);
+         _squash_removed_values(std::move(_stack.back()), head_minus_one);
```

Comment:
See @swatanabe-b1's comment about not moving more than once: https://github.com/EOSIO/chainbase/pull/51/files#commitcomment-34481085

Reply:
Resolved in [28f37d5c88015e37e687a4d79b33de440345e377](https://github.com/EOSIO/chainbase/blob/28f37d5c88015e37e687a4d79b33de440345e377/include/chainbase/chainrocks.hpp#L331-L352).